// @pixeleye/cli-logger/dist/index.js
// src/index.ts
function getLogLevel() {
  return process.env.PIXELEYE_LOG_LEVEL || "info";
}
var levels = {
  error: 0,
  warn: 1,
  info: 2,
  http: 3,
  verbose: 4,
  debug: 5,
  silly: 6
};
function setLogLevel(level) {
  process.env.PIXELEYE_LOG_LEVEL = level;
}
function shouldLog(level) {
  return levels[level] <= levels[getLogLevel()];
}
function error(message, ...optionalParams) {
  if (shouldLog("error"))
    console.error(message, ...optionalParams);
}
function warn(message, ...optionalParams) {
  if (shouldLog("warn"))
    console.warn(message, ...optionalParams);
}
function info(message, ...optionalParams) {
  if (shouldLog("info"))
    console.info(message, ...optionalParams);
}
function http(message, ...optionalParams) {
  if (shouldLog("http"))
    console.log(message, ...optionalParams);
}
function verbose(message, ...optionalParams) {
  if (shouldLog("verbose"))
    console.log(message, ...optionalParams);
}
function debug(message, ...optionalParams) {
  if (shouldLog("debug"))
    console.log(message, ...optionalParams);
}
function silly(message, ...optionalParams) {
  if (shouldLog("silly"))
    console.log(message, ...optionalParams);
}
var logger = {
  error,
  warn,
  info,
  http,
  verbose,
  debug,
  silly
};
export {
  getLogLevel,
  levels,
  logger,
  setLogLevel
};
